File: src/csharp/Json.cs
using LanguageExt;
using LanguageExt.Common;
using LanguageExt.Traits;
using System;
using System.Collections.Frozen;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;

namespace common;

public static class JsonObjectModule
{
    public static JsonResult<JsonNode> GetProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject switch
        {
            null => JsonResult.Fail<JsonNode>("JSON object is null."),
            _ => jsonObject.TryGetPropertyValue(propertyName, out var jsonNode)
                    ? jsonNode switch
                    {
                        null => JsonResult.Fail<JsonNode>($"Property '{propertyName}' is null."),
                        _ => JsonResult.Succeed(jsonNode)
                    }
                    : JsonResult.Fail<JsonNode>($"JSON object does not have a property named '{propertyName}'.")
        };

    public static Option<JsonNode> GetOptionalProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName)
                  .Match(Option<JsonNode>.Some,
                         _ => Option<JsonNode>.None);

    public static JsonResult<JsonObject> GetJsonObjectProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName,
                               jsonNode => jsonNode.AsJsonObject());

    private static JsonResult<T> GetProperty<T>(this JsonObject? jsonObject, string propertyName, Func<JsonNode, JsonResult<T>> selector) =>
        jsonObject.GetProperty(propertyName)
                  .Bind(selector)
                  .AddPropertyNameToErrorMessage(propertyName);

    private static JsonResult<T> AddPropertyNameToErrorMessage<T>(this JsonResult<T> result, string propertyName)
    {
        return result.ReplaceError(replaceError);

        JsonError replaceError(JsonError error) =>
            JsonError.From($"Property '{propertyName}' is invalid. {error.Message}");
    }

    public static JsonResult<JsonArray> GetJsonArrayProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName,
                               jsonNode => jsonNode.AsJsonArray());

    public static JsonResult<JsonValue> GetJsonValueProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName,
                               jsonNode => jsonNode.AsJsonValue());

    public static JsonResult<string> GetStringProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName,
                               jsonNode => jsonNode.AsJsonValue()
                                                   .Bind(jsonValue => jsonValue.AsString()));

    public static JsonResult<int> GetIntProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName,
                               jsonNode => jsonNode.AsJsonValue()
                                                   .Bind(jsonValue => jsonValue.AsInt()));

    public static JsonResult<bool> GetBoolProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName,
                               jsonNode => jsonNode.AsJsonValue()
                                                   .Bind(jsonValue => jsonValue.AsBool()));

    public static JsonResult<Guid> GetGuidProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName,
                               jsonNode => jsonNode.AsJsonValue()
                                                   .Bind(jsonValue => jsonValue.AsGuid()));

    public static JsonResult<Uri> GetAbsoluteUriProperty(this JsonObject? jsonObject, string propertyName) =>
        jsonObject.GetProperty(propertyName,
                               jsonNode => jsonNode.AsJsonValue()
                                                   .Bind(jsonValue => jsonValue.AsAbsoluteUri()));

    public static JsonObject SetProperty(this JsonObject jsonObject, string propertyName, JsonNode? propertyValue)
    {
        jsonObject[propertyName] = propertyValue;
        return jsonObject;
    }

    public static JsonResult<JsonObject> ToJsonObject(BinaryData? data, JsonSerializerOptions? options = default) =>
        JsonNodeModule.Deserialize<JsonObject>(data, options);
}

public static class JsonValueModule
{
    public static JsonResult<string> AsString(this JsonValue? jsonValue) =>
        jsonValue?.GetValueKind() switch
        {
            JsonValueKind.String => jsonValue.GetStringValue() switch
            {
                null => JsonResult.Fail<string>("JSON value has a null string."),
                var stringValue => JsonResult.Succeed(stringValue)
            },
            _ => JsonResult.Fail<string>("JSON value is not a string.")
        };

    private static string? GetStringValue(this JsonValue? jsonValue) =>
        jsonValue?.GetValue<object>().ToString();

    public static JsonResult<int> AsInt(this JsonValue? jsonValue)
    {
        var errorMessage = "JSON value is not an integer.";

        return jsonValue?.GetValueKind() switch
        {
            JsonValueKind.Number => int.TryParse(jsonValue.GetStringValue(), out var result)
                                    ? JsonResult.Succeed(result)
                                    : JsonResult.Fail<int>(errorMessage),
            _ => JsonResult.Fail<int>(errorMessage)
        };
    }

    public static JsonResult<bool> AsBool(this JsonValue? jsonValue) =>
        jsonValue?.GetValueKind() switch
        {
            JsonValueKind.True => JsonResult.Succeed(true),
            JsonValueKind.False => JsonResult.Succeed(false),
            _ => JsonResult.Fail<bool>("JSON value is not a boolean.")
        };

    public static JsonResult<Guid> AsGuid(this JsonValue? jsonValue)
    {
        var errorMessage = "JSON value is not a GUID.";

        return jsonValue.AsString()
                        .Bind(stringValue => Guid.TryParse(jsonValue.GetStringValue(), out var result)
                                            ? JsonResult.Succeed(result)
                                            : JsonResult.Fail<Guid>(errorMessage))
                        .ReplaceError(errorMessage);
    }

    public static JsonResult<Uri> AsAbsoluteUri(this JsonValue? jsonValue)
    {
        var errorMessage = "JSON value is not an absolute URI.";

        return jsonValue.AsString()
                 .Bind(stringValue => Uri.TryCreate(jsonValue.GetStringValue(), UriKind.Absolute, out var result)
                                        ? JsonResult.Succeed(result)
                                        : JsonResult.Fail<Uri>(errorMessage))
                 .ReplaceError(errorMessage);
    }
}

public static class JsonArrayModule
{
    public static JsonArray ToJsonArray(this IEnumerable<JsonNode?> nodes) =>
        new([.. nodes]);

    public static ValueTask<JsonArray> ToJsonArray(this IAsyncEnumerable<JsonNode?> nodes, CancellationToken cancellationToken) =>
        nodes.AggregateAsync(new JsonArray(),
                                (array, node) =>
                                {
                                    array.Add(node);
                                    return array;
                                },
                                cancellationToken);

    public static JsonResult<ImmutableArray<JsonObject>> GetJsonObjects(this JsonArray jsonArray) =>
        jsonArray.GetElements(jsonNode => jsonNode.AsJsonObject(),
                              index => JsonError.From($"Node at index {index} is not a JSON object."));

    private static JsonResult<ImmutableArray<T>> GetElements<T>(this JsonArray jsonArray, Func<JsonNode?, JsonResult<T>> selector, Func<int, JsonError> errorFromIndex)
    {
        return jsonArray.Select((node, index) => (node, index))
                        .AsIterable()
                        .Traverse(x => nodeToElement(x.node, x.index))
                        .Map(iterable => iterable.ToImmutableArray())
                        .As();

        JsonResult<T> nodeToElement(JsonNode? node, int index) =>
            selector(node)
                .ReplaceError(_ => errorFromIndex(index));
    }

    public static JsonResult<ImmutableArray<JsonArray>> GetJsonArrays(this JsonArray jsonArray) =>
        jsonArray.GetElements(jsonNode => jsonNode.AsJsonArray(),
                              index => JsonError.From($"Node at index {index} is not a JSON array."));

    public static JsonResult<ImmutableArray<JsonValue>> GetJsonValues(this JsonArray jsonArray) =>
        jsonArray.GetElements(jsonNode => jsonNode.AsJsonValue(),
                              index => JsonError.From($"Node at index {index} is not a JSON value."));
}

public static class JsonNodeModule
{
    public static JsonResult<JsonObject> AsJsonObject(this JsonNode? node) =>
        node switch
        {
            JsonObject jsonObject => JsonResult.Succeed(jsonObject),
            null => JsonResult.Fail<JsonObject>("JSON node is null."),
            _ => JsonResult.Fail<JsonObject>("JSON node is not a JSON object.")
        };

    public static JsonResult<JsonArray> AsJsonArray(this JsonNode? node) =>
        node switch
        {
            JsonArray jsonArray => JsonResult.Succeed(jsonArray),
            null => JsonResult.Fail<JsonArray>("JSON node is null."),
            _ => JsonResult.Fail<JsonArray>("JSON node is not a JSON array.")
        };

    public static JsonResult<JsonValue> AsJsonValue(this JsonNode? node) =>
        node switch
        {
            JsonValue jsonValue => JsonResult.Succeed(jsonValue),
            null => JsonResult.Fail<JsonValue>("JSON node is null."),
            _ => JsonResult.Fail<JsonValue>("JSON node is not a JSON value.")
        };

    public static async ValueTask<JsonResult<JsonNode>> From(Stream? data,
                                                             JsonNodeOptions? nodeOptions = default,
                                                             JsonDocumentOptions documentOptions = default,
                                                             CancellationToken cancellationToken = default)
    {
        try
        {
            return data switch
            {
                null => JsonResult.Fail<JsonNode>("Stream is null."),
                _ => await JsonNode.ParseAsync(data, nodeOptions, documentOptions, cancellationToken) switch
                {
                    null => JsonResult.Fail<JsonNode>("Deserialization returned a null result."),
                    var node => JsonResult.Succeed(node)
                }
            };
        }
        catch (JsonException exception)
        {
            var jsonError = JsonError.From(exception);
            return JsonResult.Fail<JsonNode>(jsonError);
        }
    }

    public static JsonResult<JsonNode> From(BinaryData? data, JsonNodeOptions? options = default)
    {
        try
        {
            return data switch
            {
                null => JsonResult.Fail<JsonNode>("Binary data is null."),
                _ => JsonNode.Parse(data, options) switch
                {
                    null => JsonResult.Fail<JsonNode>("Deserialization returned a null result."),
                    var node => JsonResult.Succeed(node)
                }
            };
        }
        catch (JsonException exception)
        {
            var jsonError = JsonError.From(exception);
            return JsonResult.Fail<JsonNode>(jsonError);
        }
    }

    public static JsonResult<T> Deserialize<T>(BinaryData? data, JsonSerializerOptions? options = default)
    {
        if (data is null)
        {
            return JsonResult.Fail<T>("Binary data is null.");
        }

        try
        {
            var jsonObject = JsonSerializer.Deserialize<T>(data, options ?? JsonSerializerOptions.Web);

            return jsonObject is null
                ? JsonResult.Fail<T>("Deserialization return a null result.")
                : JsonResult.Succeed(jsonObject);
        }
        catch (JsonException exception)
        {
            var jsonError = JsonError.From(exception);
            return JsonResult.Fail<T>(jsonError);
        }
    }

    public static Stream ToStream(JsonNode node, JsonSerializerOptions? options = default) =>
        BinaryData.FromObjectAsJson(node, options ?? JsonSerializerOptions.Web)
                  .ToStream();
}

public sealed record JsonError : Semigroup<JsonError>
{
    private readonly FrozenSet<string> messages;

    private JsonError(IEnumerable<string> messages) => this.messages = messages.ToFrozenSet();

    public static JsonError From(string message) => new([message]);

    public static JsonError From(Exception exception) =>
        exception switch
        {
            AggregateException aggregateException =>
                new JsonError([exception.Message,
                                ..aggregateException.Flatten()
                                                    .InnerExceptions
                                                    .Select(exception => exception.Message)]),
            _ => new([exception.Message])
        };

    public string Message => messages.First();

    internal FrozenSet<string> Messages => messages;

    public JsonException ToException() =>
        messages.ToArray() switch
        {
            [var message] => new JsonException(message),
            _ => new JsonException("Multiple errors, see inner exception for details.",
                                    new AggregateException(messages.Select(message => new JsonException(message))))
        };

    public JsonError Combine(JsonError rhs) =>
        new(messages.Concat(rhs.messages));

    public static JsonError operator +(JsonError lhs, JsonError rhs) =>
        lhs.Combine(rhs);
}

public class JsonResult :
    Monad<JsonResult>,
    Traversable<JsonResult>,
    Choice<JsonResult>
{
    public static JsonResult<T> Succeed<T>(T value) =>
        JsonResult<T>.Succeed(value);

    public static JsonResult<T> Fail<T>(JsonError error) =>
        JsonResult<T>.Fail(error);

    public static JsonResult<T> Fail<T>(string errorMessage) =>
        Fail<T>(JsonError.From(errorMessage));

    public static K<JsonResult, T> Pure<T>(T value) =>
        Succeed(value);

    public static K<JsonResult, T2> Bind<T1, T2>(K<JsonResult, T1> ma, Func<T1, K<JsonResult, T2>> f) =>
        ma.As()
          .Match(f, Fail<T2>);

    public static K<JsonResult, T2> Map<T1, T2>(Func<T1, T2> f, K<JsonResult, T1> ma) =>
        ma.As()
          .Match(t1 => Pure(f(t1)),
                 Fail<T2>);

    public static K<JsonResult, T2> Apply<T1, T2>(K<JsonResult, Func<T1, T2>> mf, K<JsonResult, T1> ma) =>
        mf.As()
          .Match(f => ma.Map(f),
                 error1 => ma.As()
                             .Match(t1 => Fail<T2>(error1),
                                    error2 => Fail<T2>(error1 + error2)));

    public static K<JsonResult, T> Choose<T>(K<JsonResult, T> fa, K<JsonResult, T> fb) =>
        Choose(fa, () => fb);

    public static K<JsonResult, T> Choose<T>(K<JsonResult, T> fa, Func<K<JsonResult, T>> fb) =>
        fa.As()
          .Match(_ => fa,
                 _ => fb());

    public static K<JsonResult, A> Combine<A>(K<JsonResult, A> lhs, K<JsonResult, A> rhs) =>
        lhs.Choose(rhs);

    public static K<TApplicative, K<JsonResult, T2>> Traverse<TApplicative, T1, T2>(Func<T1, K<TApplicative, T2>> f, K<JsonResult, T1> ta) where TApplicative : Applicative<TApplicative> =>
        (K<TApplicative, K<JsonResult, T2>>)
        ta.As()
          .Match(t1 => f(t1).Map(Succeed),
                 error => TApplicative.Pure(Fail<T2>(error)));

    public static S FoldWhile<A, S>(Func<A, Func<S, S>> f, Func<(S State, A Value), bool> predicate, S initialState, K<JsonResult, A> ta) =>
        ta.As()
          .Match(a => predicate((initialState, a))
                          ? f(a)(initialState)
                          : initialState,
                 _ => initialState);

    public static S FoldBackWhile<A, S>(Func<S, Func<A, S>> f, Func<(S State, A Value), bool> predicate, S initialState, K<JsonResult, A> ta) =>
        ta.As()
          .Match(a => predicate((initialState, a))
                          ? f(initialState)(a)
                          : initialState,
                 _ => initialState);

    public static JsonResult<T> FromFin<T>(Fin<T> fin) =>
        fin.Match(Succeed,
                  error => Fail<T>(error.Message));
}

public class JsonResult<T> :
    IEquatable<JsonResult<T>>,
    K<JsonResult, T>
{
    private readonly Either<JsonError, T> value;

    private JsonResult(Either<JsonError, T> value) => this.value = value;

    public T2 Match<T2>(Func<T, T2> Succ, Func<JsonError, T2> Fail) =>
        value.Match(Fail, Succ);

    public Unit Match(Action<T> Succ, Action<JsonError> Fail) =>
        value.Match(Fail, Succ);

    public JsonResult<T2> Map<T2>(Func<T, T2> f) =>
        new(value.Map(f));

    public JsonResult<T2> Bind<T2>(Func<T, JsonResult<T2>> f) =>
        new(value.Bind(t => f(t).value));

    internal static JsonResult<T> Succeed(T value) =>
        new(value);

    internal static JsonResult<T> Fail(JsonError error) =>
        new(error);

    public override bool Equals(object? obj) =>
        obj is JsonResult<T> result && Equals(result);

    public override int GetHashCode() =>
        value.GetHashCode();

    public bool Equals(JsonResult<T>? other) =>
        other is not null
        && this.Match(t => other.Match(t2 => t?.Equals(t2) ?? false,
                                       _ => false),
                      error => other.Match(_ => false,
                                           error2 => error.Equals(error2)));
}

public static class JsonResultExtensions
{
    public static JsonResult<T> As<T>(this K<JsonResult, T> k) =>
        (JsonResult<T>)k;

    public static JsonResult<T> ReplaceError<T>(this JsonResult<T> result, string newErrorMessage) =>
        result.ReplaceError(JsonError.From(newErrorMessage));

    public static JsonResult<T> ReplaceError<T>(this JsonResult<T> result, JsonError newError) =>
        result.ReplaceError(_ => newError);

    public static JsonResult<T> ReplaceError<T>(this JsonResult<T> result, Func<JsonError, JsonError> f) =>
        result.Match(_ => result,
                     error => JsonResult.Fail<T>(f(error)));

    public static T IfFail<T>(this JsonResult<T> result, Func<JsonError, T> f) =>
        result.Match(t => t, f);

    public static T? DefaultIfFail<T>(this JsonResult<T> result) =>
        result.Match<T?>(t => t, _ => default);

    public static T ThrowIfFail<T>(this K<JsonResult, T> k) =>
        k.As().ThrowIfFail();

    public static T ThrowIfFail<T>(this JsonResult<T> result) =>
        result.IfFail(error => throw error.ToException());

    public static Fin<T> ToFin<T>(this K<JsonResult, T> k) =>
        k.As().Match(Fin<T>.Succ,
                     jsonError => jsonError.Messages.ToArray() switch
                     {
                         [var message] => Error.New(message),
                         var errors => Error.Many(errors.Select(Error.New).ToArray())
                     });
}

File: src/fsharp/Json.fs
namespace common

#nowarn 3265

open System
open System.IO
open System.Text.Json
open System.Text.Json.Nodes
open FSharpPlus

type JsonError = private JsonError of Set<string>

[<RequireQualifiedAccess>]
module JsonError =
    let fromString message = Set.singleton message |> JsonError

    let fromException (exn: Exception) =
        match exn with
        | :? AggregateException as aggregateException ->
            aggregateException.Flatten().InnerExceptions
            |> Seq.map _.Message
            |> Seq.append [ aggregateException.Message ]
            |> Set.ofSeq
            |> JsonError
        | _ -> fromString exn.Message

    let getMessage (JsonError errors) = Seq.head errors

    let getMessages (JsonError errors) = errors

    let toJsonException (JsonError error) =
        match List.ofSeq error with
        | [ message ] -> new JsonException(message)
        | messages ->
            let aggregateException =
                messages
                |> Seq.map (fun message -> new JsonException(message) :> Exception)
                |> AggregateException

            new JsonException("Multiple errors, see inner exception for details..", aggregateException)

type JsonError with
    // Semigroup
    static member (+)(JsonError existingMessages, JsonError newMessages) =
        Set.union newMessages existingMessages |> JsonError

type JsonResult<'a> =
    | Success of 'a
    | Failure of JsonError

[<RequireQualifiedAccess>]
module JsonResult =
    let succeed x = Success x

    let fail e = Failure e

    let failWithMessage message = JsonError.fromString message |> fail

    let replaceErrorWith f jsonResult =
        match jsonResult with
        | Failure e -> f e |> fail
        | _ -> jsonResult

    let setErrorMessage message jsonResult =
        replaceErrorWith (fun _ -> JsonError.fromString message) jsonResult

    let map f jsonResult =
        match jsonResult with
        | Success x -> succeed (f x)
        | Failure e -> fail e

    let apply f x =
        match f, x with
        | Success f, Success x -> succeed (f x)
        | Failure e, Success _ -> fail e
        | Success _, Failure e -> fail e
        | Failure e1, Failure e2 -> fail (e1 + e2)

    let defaultWith f jsonResult =
        match jsonResult with
        | Success x -> x
        | Failure jsonError -> f jsonError

    let throwIfFail jsonResult =
        let throw error =
            JsonError.toJsonException error |> raise

        jsonResult |> defaultWith throw

type JsonResult<'a> with
    // Functor
    static member Map(x, f) =
        match x with
        | Success x -> f x |> JsonResult.succeed
        | Failure e -> JsonResult.fail e

    static member Unzip x =
        match x with
        | Success(x, y) -> JsonResult.succeed x, JsonResult.succeed y
        | Failure e -> JsonResult.fail e, JsonResult.fail e

    // Applicative
    static member Return x = JsonResult.succeed x

    static member (<*>)(f, x) = JsonResult.apply f x

    static member Lift2(f, x1, x2) =
        match x1, x2 with
        | Success x1, Success x2 -> f x1 x2 |> JsonResult.succeed
        | Failure e, Success _ -> JsonResult.fail e
        | Success _, Failure e -> JsonResult.fail e
        | Failure e1, Failure e2 -> JsonResult.fail (e1 + e2)

    static member Lift3(f, x1, x2, x3) =
        match x1, x2, x3 with
        | Success x1, Success x2, Success x3 -> f x1 x2 x3 |> JsonResult.succeed
        | Failure e1, Success _, Success _ -> JsonResult.fail e1
        | Failure e1, Success _, Failure e3 -> JsonResult.fail (e1 + e3)
        | Failure e1, Failure e2, Success _ -> JsonResult.fail (e1 + e2)
        | Failure e1, Failure e2, Failure e3 -> JsonResult.fail (e1 + e2 + e3)
        | Success _, Failure e2, Success _ -> JsonResult.fail e2
        | Success _, Failure e2, Failure e3 -> JsonResult.fail (e2 + e3)
        | Success _, Success _, Failure e3 -> JsonResult.fail e3

    // Zip applicative
    static member Pure x = JsonResult.succeed x

    static member (<.>)(f, x) = JsonResult.apply f x

    static member Zip(x1, x2) =
        match (x1, x2) with
        | Success x1, Success x2 -> JsonResult.succeed (x1, x2)
        | Failure e1, Success _ -> JsonResult.fail e1
        | Success _, Failure e2 -> JsonResult.fail e2
        | Failure e1, Failure e2 -> JsonResult.fail (e1 + e2)

    static member Map2(f, x1, x2) =
        match x1, x2 with
        | Success x1, Success x2 -> f x1 x2 |> JsonResult.succeed
        | Failure e, Success _ -> JsonResult.fail e
        | Success _, Failure e -> JsonResult.fail e
        | Failure e1, Failure e2 -> JsonResult.fail (e1 + e2)

    static member Map3(f, x1, x2, x3) =
        match x1, x2, x3 with
        | Success x1, Success x2, Success x3 -> f x1 x2 x3 |> JsonResult.succeed
        | Failure e1, Success _, Success _ -> JsonResult.fail e1
        | Failure e1, Success _, Failure e3 -> JsonResult.fail (e1 + e3)
        | Failure e1, Failure e2, Success _ -> JsonResult.fail (e1 + e2)
        | Failure e1, Failure e2, Failure e3 -> JsonResult.fail (e1 + e2 + e3)
        | Success _, Failure e2, Success _ -> JsonResult.fail e2
        | Success _, Failure e2, Failure e3 -> JsonResult.fail (e2 + e3)
        | Success _, Success _, Failure e3 -> JsonResult.fail e3

    // Monad
    static member Bind(x, f) =
        match x with
        | Success x -> f x
        | Failure e -> JsonResult.fail e

    static member (>>=)(x, f) = JsonResult<'a>.Bind(x, f)

    static member Join x =
        match x with
        | Success(Success x) -> JsonResult.succeed x
        | Success(Failure e) -> JsonResult.fail e
        | Failure e -> JsonResult.fail e

    // Foldable
    static member ToSeq x =
        match x with
        | Success x -> Seq.singleton x
        | Failure _ -> Seq.empty


[<RequireQualifiedAccess>]
module JsonNode =
    let asJsonObject (node: JsonNode | null) =
        match node with
        | :? JsonObject as jsonObject -> JsonResult.succeed jsonObject
        | Null -> JsonResult.failWithMessage "JSON node is null"
        | _ -> JsonResult.failWithMessage "JSON node is not a JSON object"

    let asJsonArray (node: JsonNode | null) =
        match node with
        | :? JsonArray as jsonArray -> JsonResult.succeed jsonArray
        | Null -> JsonResult.failWithMessage "JSON node is null"
        | _ -> JsonResult.failWithMessage "JSON node is not a JSON array"

    let asJsonValue (node: JsonNode | null) =
        match node with
        | :? JsonValue as jsonValue -> JsonResult.succeed jsonValue
        | Null -> JsonResult.failWithMessage "JSON node is null"
        | _ -> JsonResult.failWithMessage "JSON node is not a JSON value"

    let private ifNoneNullable option =
        option |> Option.map Nullable |> Option.defaultValue (Nullable())

    let private ifNoneDefault option =
        option |> Option.defaultValue Unchecked.defaultof<'a>

    let fromStreamWithOptions (stream: Stream | null) nodeOptions documentOptions =
        async {
            let! cancellationToken = Async.CancellationToken

            match stream with
            | Null -> return JsonResult.failWithMessage "Stream is null."
            | NonNull stream ->
                try
                    let nodeOptions = ifNoneNullable nodeOptions
                    let documentOptions = ifNoneDefault documentOptions

                    match!
                        JsonNode.ParseAsync(stream, nodeOptions, documentOptions, cancellationToken = cancellationToken)
                        |> Async.AwaitTask
                    with
                    | Null -> return JsonResult.failWithMessage "Deserialization returned a null result."
                    | NonNull node -> return JsonResult.succeed node
                with exn ->
                    return JsonError.fromException exn |> JsonResult.fail
        }

    let fromStream stream = fromStreamWithOptions stream None None

    let fromBinaryDataWithOptions (data: BinaryData | null) nodeOptions =
        try
            match data with
            | Null -> JsonResult.failWithMessage "Binary data is null."
            | NonNull data ->
                let nodeOptions = ifNoneNullable nodeOptions

                match JsonNode.Parse(data, nodeOptions) with
                | Null -> JsonResult.failWithMessage "Deserialization returned a null result."
                | NonNull node -> JsonResult.succeed node
        with exn ->
            JsonError.fromException exn |> JsonResult.fail

    let fromBinaryData data = fromBinaryDataWithOptions data None

    let toBinaryData (node: JsonNode) = BinaryData.FromObjectAsJson(node)

    let toStream (node: JsonNode) = toBinaryData node |> _.ToStream()

[<RequireQualifiedAccess>]
module JsonArray =
    let private toSeq (jsonArray: JsonArray) = jsonArray |> seq

    let fromSeq seq = Array.ofSeq seq |> JsonArray

    let private getResultSeq toJsonResult toErrorMessage (jsonArray: JsonArray) =
        jsonArray
        |> toSeq
        |> traversei (fun index node ->
            let replaceErrorMessage = toErrorMessage index |> JsonResult.setErrorMessage
            toJsonResult node |> replaceErrorMessage)

    let asJsonObjects jsonArray =
        getResultSeq JsonNode.asJsonObject (fun index -> $"Element at index {index} is not a JSON object.") jsonArray

    let asJsonArrays jsonArray =
        getResultSeq JsonNode.asJsonArray (fun index -> $"Element at index {index} is not a JSON array.") jsonArray

    let asJsonValues jsonArray =
        getResultSeq JsonNode.asJsonValue (fun index -> $"Element at index {index} is not a JSON value.") jsonArray

    let getJsonObjects jsonArray =
        jsonArray
        |> toSeq
        |> Seq.choose (fun node ->
            match node with
            | :? JsonObject as jsonObject -> Some jsonObject
            | _ -> None)

    let getJsonArrays jsonArray =
        jsonArray
        |> toSeq
        |> Seq.choose (fun node ->
            match node with
            | :? JsonArray as jsonArray -> Some jsonArray
            | _ -> None)

    let getJsonValues jsonArray =
        jsonArray
        |> toSeq
        |> Seq.choose (fun node ->
            match node with
            | :? JsonValue as jsonValue -> Some jsonValue
            | _ -> None)

[<RequireQualifiedAccess>]
module JsonValue =
    let private getString (jsonValue: JsonValue) = jsonValue.GetValue<obj>() |> string

    let asString (jsonValue: JsonValue) =
        match jsonValue.GetValueKind() with
        | JsonValueKind.String -> getString jsonValue |> JsonResult.succeed
        | _ -> JsonResult.failWithMessage "JSON value is not a string"

    let asInt (jsonValue: JsonValue) =
        match jsonValue.GetValueKind() with
        | JsonValueKind.Number ->
            match getString jsonValue |> Int32.TryParse with
            | true, x -> JsonResult.succeed x
            | _ -> JsonResult.failWithMessage "JSON value is not an integer"
        | _ -> JsonResult.failWithMessage "JSON value is not a number"

    let asAbsoluteUri jsonValue =
        let errorMessage = "JSON value is not an absolute URI."

        asString jsonValue
        |> bind (fun stringValue ->
            match Uri.TryCreate(stringValue, UriKind.Absolute) with
            | true, uri when
                match uri with
                | Null -> false
                | NonNull nonNullUri -> nonNullUri.HostNameType <> UriHostNameType.Unknown
                ->
                JsonResult.succeed uri
            | _ -> JsonResult.failWithMessage errorMessage)
        |> JsonResult.setErrorMessage errorMessage

    let asGuid jsonValue =
        asString jsonValue
        |> bind (fun stringValue ->
            match Guid.TryParse(stringValue) with
            | true, guid -> JsonResult.succeed guid
            | _ -> JsonResult.failWithMessage "JSON value is not a GUID.")

    let asBool (jsonValue: JsonValue) =
        match jsonValue.GetValueKind() with
        | JsonValueKind.True -> JsonResult.succeed true
        | JsonValueKind.False -> JsonResult.succeed false
        | _ -> JsonResult.failWithMessage "JSON value is not a boolean."

    let asDateTimeOffset (jsonValue: JsonValue) =
        let errorMessage = "JSON value is not a date time offset."

        match jsonValue.TryGetValue<DateTimeOffset>() with
        | true, dateTimeOffset -> JsonResult.succeed dateTimeOffset
        | _ ->
            monad {
                let! stringValue = asString jsonValue |> JsonResult.setErrorMessage errorMessage

                match DateTimeOffset.TryParse(stringValue) with
                | true, dateTimeOffset -> return dateTimeOffset
                | _ -> return! JsonResult.failWithMessage errorMessage
            }

[<RequireQualifiedAccess>]
module JsonObject =
    let getProperty propertyName (jsonObject: JsonObject | null) =
        match jsonObject with
        | null -> JsonResult.failWithMessage "JSON object is null."
        | jsonObject ->
            match jsonObject.TryGetPropertyValue(propertyName) with
            | true, property ->
                match property with
                | null -> JsonResult.failWithMessage $"Property '{propertyName}' is null."
                | property -> JsonResult.succeed property
            | _ -> JsonResult.failWithMessage $"Property '{propertyName}' is missing."

    let getOptionalProperty propertyName jsonObject =
        getProperty propertyName jsonObject
        |> map Option.Some
        |> JsonResult.defaultWith (fun _ -> Option.None)

    let private addPropertyNameToErrorMessage propertyName jsonResult =
        let replaceError jsonError =
            let originalErrorMessage = JsonError.getMessage jsonError

            let newErrorMessage =
                $"Property '{propertyName}' is invalid. {originalErrorMessage}"

            JsonError.fromString newErrorMessage

        JsonResult.replaceErrorWith replaceError jsonResult

    let private getPropertyFromResult getPropertyResult propertyName jsonObject =
        getProperty propertyName jsonObject
        |> bind getPropertyResult
        |> addPropertyNameToErrorMessage propertyName

    let getJsonObjectProperty propertyName jsonObject =
        getPropertyFromResult JsonNode.asJsonObject propertyName jsonObject

    let getJsonArrayProperty propertyName jsonObject =
        getPropertyFromResult JsonNode.asJsonArray propertyName jsonObject

    let getJsonValueProperty propertyName jsonObject =
        getPropertyFromResult JsonNode.asJsonValue propertyName jsonObject

    let getStringProperty propertyName jsonObject =
        let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asString
        getPropertyFromResult getPropertyResult propertyName jsonObject

    let getAbsoluteUriProperty propertyName jsonObject =
        let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asAbsoluteUri
        getPropertyFromResult getPropertyResult propertyName jsonObject

    let getGuidProperty propertyName jsonObject =
        let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asGuid
        getPropertyFromResult getPropertyResult propertyName jsonObject

    let getBoolProperty propertyName jsonObject =
        let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asBool
        getPropertyFromResult getPropertyResult propertyName jsonObject

    let getIntProperty propertyName jsonObject =
        let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asInt
        getPropertyFromResult getPropertyResult propertyName jsonObject

    let getDateTimeOffsetProperty propertyName jsonObject =
        let getPropertyResult = JsonNode.asJsonValue >> bind JsonValue.asDateTimeOffset
        getPropertyFromResult getPropertyResult propertyName jsonObject

    let setProperty (propertyName: string) (propertyValue: JsonNode | null) (jsonObject: JsonObject) =
        jsonObject[propertyName] <- propertyValue
        jsonObject

    let removeProperty (propertyName: string) (jsonObject: JsonObject) =
        jsonObject.Remove(propertyName) |> ignore
        jsonObject

    let fromStreamWithOptions stream nodeOptions documentOptions =
        JsonNode.fromStreamWithOptions stream nodeOptions documentOptions
        |> map (bind JsonNode.asJsonObject)

    let fromStream stream = fromStreamWithOptions stream None None

    let fromBinaryDataWithOptions data nodeOptions =
        JsonNode.fromBinaryDataWithOptions data nodeOptions
        |> bind JsonNode.asJsonObject

    let fromBinaryData data = fromBinaryDataWithOptions data None
